packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,462 @@
+
+    if (connectionRecord.invitation?.id !== message.thread?.parentThreadId) {
+      throw new ProblemReportError('Missing reference to invitation.', { problemCode: 'request_not_accepted' })
+    }
+
+    await this.updateState(DidExchangeCompleteMessage.type, connectionRecord)
+    this.logger.debug(`Process message ${DidExchangeCompleteMessage.type} end`, { connectionRecord })
+    return connectionRecord
+  }
+
+  private async updateState(messageType: string, connectionRecord: ConnectionRecord) {
+    this.logger.debug(`Updating state`, { connectionRecord })
+    const nextState = DidExchangeStateMachine.nextState(messageType, connectionRecord)
+    return this.connectionService.updateState(connectionRecord, nextState)
+  }
+
+  private async createPeerDidDoc(routing: Routing) {
+    const publicKeyBase58 = routing.verkey
+
+    const ed25519Key = Key.fromPublicKeyBase58(publicKeyBase58, KeyType.Ed25519)
+    const x25519Key = Key.fromPublicKey(convertPublicKeyToX25519(ed25519Key.publicKey), KeyType.X25519)
+
+    // For peer dids generated with method 1, the controller MUST be #id as we don't know the did yet
+    const ed25519VerificationMethod = getEd25519VerificationMethod({
+      id: uuid(),
+      key: ed25519Key,
+      controller: '#id',
+    })
+    const x25519VerificationMethod = getX25519VerificationMethod({
+      id: uuid(),
+      key: x25519Key,
+      controller: '#id',
+    })
+
+    let mediatorRoutingKey
+    if (routing.routingKeys.length > 0) {
+      const [mediatorPublicKeyBase58] = routing.routingKeys
+      const mediatorEd25519Key = Key.fromPublicKeyBase58(mediatorPublicKeyBase58, KeyType.Ed25519)
+      const mediatorEd25519DidKey = new DidKey(mediatorEd25519Key)
+      const mediatorX25519Key = Key.fromPublicKey(
+        convertPublicKeyToX25519(mediatorEd25519Key.publicKey),
+        KeyType.X25519
+      )
+      // Use ed25519 did:key, which also includes the x25519 key used for didcomm
+      mediatorRoutingKey = `${mediatorEd25519DidKey.did}#${mediatorX25519Key.fingerprint}`
+    }
+
+    // TODO Iterate over all endpoints
+    const [serviceEndpoint] = routing.endpoints
+    const service = new DidCommService({
+      id: '#service-0',
+      // Fixme: can we use relative reference (#id) instead of absolute reference here (did:example:123#id)?
+      // We don't know the did yet
+      // TODO we should perhaps use keyAgreement instead of authentication key in here, then it must be changed also in connection record verkey
+      recipientKeys: [ed25519Key.publicKeyBase58],
+      serviceEndpoint,
+      accept: ['didcomm/aip2;env=rfc19'],
+      // It is important that we encode the routing keys as key references.
+      // So instead of using plain verkeys, we should encode them as did:key dids
+      routingKeys: mediatorRoutingKey ? [mediatorRoutingKey] : [],
+    })
+
+    const didDocument = new DidDocumentBuilder('')
```

Comment:
We should remove the required id property from the did document builder, it's ugly to pass an empty string. I'll make sure to create a PR for that.